### PR TITLE
Setup testing env

### PR DESCRIPTION
**Pull request by benjpalmer**
Alright.. So I made some large changes to the file structure for the application. This was for a couple reasons. First to have a more organized file structure. Second so we could import the flask application elsewhere in our code, in this case so we could create an instance of MPV and then run tests. In the process, I also refactored the existing db_setup file to import database settings from config.py instead of the depreciated use of config.json.
If you have any questions please ask!

- [ ] **Does this feature have any breaking changes?**
**YES** .It is possible this new file structure might effect @zachtheclimber production environment. If you have manually set the `FLASK_APP` ENV variable. When `flask run` is called, by default it looks for a directory named `app`, so I just followed that design pattern. That change is the most likely to effect production.  In development, everything should work as it did before, any commands have been updated in the README. 

- [ ] **Does this change greatly affect the UI/frontend of the application?**
**No**

To run the tests make sure you are in the root directory of your project and simply run the shell command `pytest`

**Where do we go from here?**

We can go from here and add tests for other things as you see fit, for example, the helper functions. We could test those. I did not elect to keep track of code coverage but if you decide you want to we can address that as well. 

I did massage all of the files and structure of the application. I hope I didn't break anything. I tested as much of the app as I could on my machine and everything seems all good, just let me know if something seems off!

### CODE REVIEW

**Comment by ghost** [comment]
Awesome! Thanks for getting this set up. I'll have to dig in to Pytest and learn how to implement more testing for existing/future code. Also, thanks for catching that the fact that I forgot to update db_setup.py with the new config scheme. :-)

File: docker-compose.yaml [hunk 1]
```diff
@@ -4,7 +4,7 @@ services:
     image: mysql:5.7
     restart: always
     environment:
-      MYSQL_DATABASE: 'mpv'
+      MYSQL_DATABASE: 'app'
```
Comment (ghost) [inline docker-compose.yaml:7]: I'm curious: Why the change in db name for the Docker setup?

File: app/__init__.py [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+import re
+
+from flask import Flask, render_template, request, redirect
+from .config import *
+
+from .helpers import get_userid, ticklist, db_load, db_connect, db_close
+from .graphing import height_climbed, pitches_climbed, grade_scatter, get_types
+
+
+def create_app(test_config=None):
```
Comment (ghost) [inline app/__init__.py:17]: This restructure kind of borked my production configuration. I can get MPV to run (both in my dev environment and on the production server) if I manually start Flask (as you noted with `flask run`). Unfortunately, when I try to run it using gunicorn (after changing my wsgi config file), it thinks I'm trying to pass values into `create_app()` rather than `app.run()`.

I spent a bunch of time today messing with the wsgi config, gunicorn, and looking for other flask implementations that use a similar layout, but unfortunately, everything I could find has `app = Flask(__name__)` outside of a function.

I figured out a workaround where I just removed  the `create_app()` def and the `test_config` if/else on the server code. I'd rather not have to do that every time I update the code there, but for now, I can't figure out a better way. If you have any thoughts, I'd love to hear them.

I'll probably try to dig into this another day, when I have time. I'm not saying your code is bad or that I want to revert, just that it is a bit hard to implement with gunicorn. :-) Anyway, no biggie. Just learning how to communicate and work remotely with someone. Thanks again for helping out with this project.

**Review by ghost — COMMENTED** [review]
A couple of thoughts. Nothing broken, as far as I can tell.

File: docker-compose.yaml (hunk 1 above)
Comment (benjpalmer) [reply docker-compose.yaml:7]: Okay that was totally my bad! I didn't intend to change that, when I was moving files around my IDE was also refactoring file paths and such, I must have accidentally changed that!

File: app/__init__.py (hunk 2 above)
Comment (benjpalmer) [reply app/__init__.py:17]: Yeah I had suspected this might effect your production environment, sorry that ended up being the case but I'm glad you were able to find a work around! I did a bit of searching for running an app factory and configuring for deployment. I'm not sure this is much help but I found this resource:

This one is gunicorn specific
https://stackoverflow.com/questions/25319690/how-do-i-run-a-flask-app-in-gunicorn-if-i-used-the-application-factory-pattern

I think this was apache specific but the logic looked to be the same.
https://stackoverflow.com/questions/36499982/how-to-deploy-a-flask-app-with-factory-function-for-application-creation

The common thread seems to be that you need to point your configuration file to an application object, and that application object can be a call to `create_app()`. 

Also from stackoverflow:

According to the Flask Documentation, an application factory is good because:

- Testing. You can have instances of the application with different settings to test every case.

- Multiple instances. Imagine you want to run different versions of the same application. Of course you could have multiple instances with different configs set up in your webserver, but if you use factories, you can have multiple instances of the same application running in the same application process which can be handy

Ultimately this was the reason for adding an application factory. So we could test against the actual MPV app.

Let me know if I can be of any help answering any other questions!

File: docker-compose.yaml (hunk 1 above)
Comment (ghost) [reply docker-compose.yaml:7]: Haha, no worries. I'll get it changed back over. :-)

File: app/__init__.py (hunk 2 above)
Comment (ghost) [reply app/__init__.py:17]: Perfect! The gunicorn specific link had just what I needed. I didn't know this was an app factory setup, so that was what was throwing me off knowing what to search for. Such an easy fix once I had the right info. Thanks for passing it along!
